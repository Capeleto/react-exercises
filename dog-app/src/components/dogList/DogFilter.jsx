import React from "react";
import { FormControlLabel, Radio, RadioGroup } from "@material-ui/core";

function DogeFilter({ data = [], value, onChange }) {
  const alpha = Array.from(Array(26)).map((e, i) => i + 65);
  const alphabet = alpha.map((x) => String.fromCharCode(x));

  const letterData = alphabet.reduce((accumulator, letter) => {
    accumulator[letter] = data.filter((element) => element.toUpperCase().startsWith(letter));

    return accumulator;
  }, {});

  function handleChange(_, letter) {
    onChange(letterData[letter]);
  }

  return (
    <RadioGroup
      aria-label="dogeFilter"
      name="dogeFilter"
      value={value}
      onChange={handleChange}
      className="radio-group"
    >
      {alphabet.map((letter) => (
        <FormControlLabel
          value={letter}
          control={<Radio />}
          label={`${letter} ${letterData[letter].length}`}
          className="radios"
        />
      ))}
    </RadioGroup>
  );
}

export default DogeFilter;
