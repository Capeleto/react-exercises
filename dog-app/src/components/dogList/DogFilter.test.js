import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import "@testing-library/jest-dom/extend-expect";

import DogFilter from "./DogFilter";

describe("should render component", () => {
  const mockFn = jest.fn();

  describe("without any prop", () =>
    exhaustiveComponentTests(<DogFilter />, true));

  describe("with single props", () => {
    exhaustiveComponentTests(<DogFilter id="id1" />);
    exhaustiveComponentTests(<DogFilter onChange={mockFn} />);
    exhaustiveComponentTests(
      <DogFilter
        data={Object.keys({
          affenpinscher: [],
          african: [],
          airedale: [],
          akita: [],
        })}
      />
    );
  });

  describe("with multiple props", () => {
    it("should render all radio buttons", () => {
      const wrapper = shallow(
        <DogFilter
          data={Object.keys({
            affenpinscher: [],
            african: [],
            airedale: [],
            akita: [],
          })}
        />
      );

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it("should execute onChange functions", () => {
      const mockFn = jest.fn().mockResolvedValue();

      const wrapper = shallow(<DogFilter onChange={mockFn} />);

      return wrapper
        .find("#dogeFilter")
        .invoke("onChange")()
        .then(() => {
          expect(mockFn).toBeCalled();
        });
    });

    it("should have multiple(26) radio buttons", () => {
        const wrapper = shallow(<DogFilter onChange={mockFn} />);

        expect(wrapper.find('.radios')).toHaveLength(26);
    })
  });
});

function exhaustiveComponentTests(component, withJunk = false) {
  it("should render the component without breaking it", () => {
    ReactDOM.render(component, document.createElement("div"));
  });

  if (withJunk) {
    it("should render the component without breaking it when junk is passed to props", () => {
      ReactDOM.render(
        React.cloneElement(component, { junk1: 12, test: "asd", junk: true }),
        document.createElement("div")
      );
    });
  }
}
