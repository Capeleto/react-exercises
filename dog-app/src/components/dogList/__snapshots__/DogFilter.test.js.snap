// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component with multiple props should render all radio buttons 1`] = `
<ForwardRef(RadioGroup)
  aria-label="dogeFilter"
  className="radio-group"
  id="dogeFilter"
  name="dogeFilter"
  onChange={[Function]}
>
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="A"
    key="A"
    label="A 4"
    value="A"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="B"
    key="B"
    label="B 0"
    value="B"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="C"
    key="C"
    label="C 0"
    value="C"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="D"
    key="D"
    label="D 0"
    value="D"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="E"
    key="E"
    label="E 0"
    value="E"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="F"
    key="F"
    label="F 0"
    value="F"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="G"
    key="G"
    label="G 0"
    value="G"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="H"
    key="H"
    label="H 0"
    value="H"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="I"
    key="I"
    label="I 0"
    value="I"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="J"
    key="J"
    label="J 0"
    value="J"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="K"
    key="K"
    label="K 0"
    value="K"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="L"
    key="L"
    label="L 0"
    value="L"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="M"
    key="M"
    label="M 0"
    value="M"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="N"
    key="N"
    label="N 0"
    value="N"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="O"
    key="O"
    label="O 0"
    value="O"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="P"
    key="P"
    label="P 0"
    value="P"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="Q"
    key="Q"
    label="Q 0"
    value="Q"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="R"
    key="R"
    label="R 0"
    value="R"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="S"
    key="S"
    label="S 0"
    value="S"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="T"
    key="T"
    label="T 0"
    value="T"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="U"
    key="U"
    label="U 0"
    value="U"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="V"
    key="V"
    label="V 0"
    value="V"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="W"
    key="W"
    label="W 0"
    value="W"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="X"
    key="X"
    label="X 0"
    value="X"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="Y"
    key="Y"
    label="Y 0"
    value="Y"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    className="radios"
    control={<WithStyles(ForwardRef(Radio)) />}
    id="Z"
    key="Z"
    label="Z 0"
    value="Z"
  />
</ForwardRef(RadioGroup)>
`;
